### PR TITLE
pam_userdb: enable GDBM support

### PR DESCRIPTION
**Pull request by ikerexxe**
* configure.ac: add `gdbm` option to `enable-db`
* modules/pam_userdb/pam_userdb.c: conditionally provide database access depending on the database technology

### CODE REVIEW

File: modules/pam_userdb/pam_userdb.c [hunk 1]
```diff
@@ -129,6 +131,76 @@ _pam_parse (pam_handle_t *pamh, int argc, const char **argv,
...
+    return key;
+}
+
+static datum
+db_nextkey(void *dbm, datum key UNUSED)
+{
+    datum nextkey;
+
+#ifdef HAVE_GDBM_H
+    nextkey = gdbm_nextkey(dbm, key);
```
Comment (ldv-alt) [inline modules/pam_userdb/pam_userdb.c:171]: Sorry, how can `key` be `UNUSED` if it is passed down to `gdbm_nextkey`?

Comment (ikerexxe) [reply modules/pam_userdb/pam_userdb.c:171]: It's unused for the non-gdbm option, where `key` isn't used. If I wouldn't have set the `UNUSED` keyword, then the CI would have failed. However, I'm open to other solutions.

Comment (ldv-alt) [reply modules/pam_userdb/pam_userdb.c:171]: This means that the use of `UNUSED` keyword also has to be conditional.

Comment (ikerexxe) [reply modules/pam_userdb/pam_userdb.c]: Done

File: modules/pam_userdb/pam_userdb.c [hunk 2]
```diff
@@ -129,6 +137,76 @@ _pam_parse (pam_handle_t *pamh, int argc, const char **argv,
...
+    datum key;
+
+#ifdef HAVE_GDBM_H
+    key = gdbm_firstkey(dbm);
+#else
+    key = dbm_firstkey(dbm);
+#endif /* HAVE_GDBM_H */
+
+    return key;
+}
```
Comment (ldv-alt) [inline modules/pam_userdb/pam_userdb.c:169]: In all these new wrapper functions, do they really need that level of complexity?
For example, `db_firstkey` could be simplified to

```
static datum
db_firstkey(void *dbm)
{
#ifdef HAVE_GDBM_H
    return gdbm_firstkey(dbm);
#else
    return dbm_firstkey(dbm);
#endif
}
```

Comment (ikerexxe) [reply modules/pam_userdb/pam_userdb.c]: Updated

File: modules/pam_userdb/pam_userdb.c [hunk 3]
```diff
@@ -129,6 +137,62 @@ _pam_parse (pam_handle_t *pamh, int argc, const char **argv,
...
+db_open(const char *database, mode_t file_mode)
+{
+#ifdef HAVE_GDBM_H
+    void *dbm = gdbm_open(database, 4096, GDBM_READER, file_mode, NULL);
+#else
+    void *dbm = dbm_open(database, O_RDONLY, file_mode);
+#endif /* HAVE_GDBM_H */
+
+    return dbm;
+}
```
Comment (ldv-alt) [inline modules/pam_userdb/pam_userdb.c:153]: Do you really need `dbm` variable here?

Comment (ikerexxe) [reply modules/pam_userdb/pam_userdb.c]: Not at all. Thanks for catching that.

**Review by ldv-alt — APPROVED** [review]

**Pull request merged**